src/Tests/Tests/Shared/TestQuaternion.cs

```diff
@@ -380,5 +380,85 @@ public static void quaternion_angle()
+            var m = new float3x3(expectedQuaternion);
+            m = mul(m, float3x3.Scale(5.18f));
+            var actualQuaternion = rotation(m);
+            TestUtils.AreEqual(0.0f, angle(actualQuaternion, expectedQuaternion) % PI * 2.0f, tolerance);
+        }
+
+        [TestCompiler]
+        public static void quaternion_rotation_from_3x3_with_nonuniform_scale()
+        {
+            const float tolerance = 1e-5f;
+            var m = new float3x3(
+                2.0f, 0.0f, 0.0f,
+                0.0f, 1.5f, 0.0f,
+                0.0f, 0.0f, 3.1f);
+            var q = rotation(m);
+            TestUtils.AreEqual(quaternion.identity, q, tolerance);
+        }
+
+        [TestCompiler]
+        public static void quaternion_rotation_from_3x3_with_negative_scale_x()
+        {
+            const float tolerance = 1e-5f;
+            var random = new Random(281u);
+            var expectedQuaternion = random.NextQuaternionRotation();
+            var m = new float3x3(expectedQuaternion);
+            m = mul(m, float3x3.Scale(-1.5f, 0.2f, 51.281f));
+            var actualQuaternion = rotation(m);
+            TestUtils.AreEqual(0.0f, angle(actualQuaternion, expectedQuaternion) % PI * 2.0f, tolerance);
+        }
+
+        [TestCompiler]
+        public static void quaternion_rotation_from_3x3_with_negative_scale_y()
+        {
+            const float tolerance = 1e-5f;
+            var random = new Random(10000u);
+            var expectedQuaternion = random.NextQuaternionRotation();
+            var m = new float3x3(expectedQuaternion);
+            m = mul(m, float3x3.Scale(1.5f, -0.2f, 51.281f));
+            var actualQuaternion = rotation(m);
+            TestUtils.AreEqual(0.0f, angle(actualQuaternion, expectedQuaternion) % PI * 2.0f, tolerance);
+        }
+
+        [TestCompiler]
+        public static void quaternion_rotation_from_3x3_with_negative_scale_z()
+        {
+            const float tolerance = 1e-5f;
+            var random = new Random(9891712u);
+            var expectedQuaternion = random.NextQuaternionRotation();
+            var m = new float3x3(expectedQuaternion);
+            m = mul(m, float3x3.Scale(1.5f, 0.2f, -51.281f));
+            var actualQuaternion = rotation(m);
+            TestUtils.AreEqual(0.0f, angle(actualQuaternion, expectedQuaternion) % PI * 2.0f, tolerance);
+        }
+
+        [TestCompiler]
+        public static void quaternion_rotation_from_3x3_with_negative_nonuniform_scale()
+        {
+            const float tolerance = 1e-5f;
+            var random = new Random(912u);
+            var expectedQuaternion = random.NextQuaternionRotation();
+            var m = new float3x3(expectedQuaternion);
+            m = mul(m, float3x3.Scale(-1.5f, -0.2f, -51.281f));
+            var actualQuaternion = rotation(m);
+            TestUtils.AreEqual(0.0f, angle(actualQuaternion, expectedQuaternion) % PI * 2.0f, tolerance);
+        }
```

Comment:
we should also add two more tests:

one with very small scale (1e-12f, 1e-12f, 1e-12f)
and one with zero scale (1,1,0)

Reply:
Sigh, I thought the tiny scale test I added was fine but there's actually a difference between the results when I run the test directly from the solution vs running the test with Unity. Unity generates NaNs during one of the jacobi iterations, so I'm digging into that now.

Reply:
I see the difference now. The existing picoscale test (https://github.cds.internal.unity3d.com/unity/dots/blob/5e6042e3d299482f6f26defac24350ac1f43db61/Packages/com.unity.animation/Unity.Animation.Tests/AffineTransformTests.cs#L58) actually uses `mathex.inverse` from Unity.Animation which rescales the input matrix. But the rotation method uses `svdRotation` which does no such thing and eventually the code runs into a normalization which causes the tiny values to underflow.

<img width="1180" alt="image" src="https://user-images.githubusercontent.com/3121968/150593365-0bd3c7ce-64aa-4f0f-bf9d-019a61c29089.png">

Do we need to handle these tiny scales in `svdRotation` to make this test pass?

Reply:
you're right, it the affine transform from UnityEngine that does have a test with pico scale, not the sdvRotation.

So we are on par with what we have in Unity so we can drop the pico scale test on svd.